### ISSUE TITLE
Libupnp error with Gerbera/DD-WRT/Entware 

### ISSUE THREAD

**Issue by robolaf** [mrjimenez/pupnp#110]
Hi, 
I am refering to this bug:
https://sourceforge.net/p/pupnp/bugs/138/

I also found the bug meanwhile (at least it works for me).
See upnpapi.c at line 3574:

`	/* Cycle through the list of interfaces looking for IP addresses.  */
	for (i = (size_t)0; i < (size_t)ifConf.ifc_len;) {
		struct ifreq *pifReq =
		    (struct ifreq *)((caddr_t) ifConf.ifc_req + i);
//		i += sizeof *pifReq; <<<<<<< removed
		i += sizeof ifConf.ifc_req; <<< insert
		/* See if this is the sort of interface we want to deal with. */
		memset(ifReq.ifr_name, 0, sizeof(ifReq.ifr_name));
		strncpy(ifReq.ifr_name, pifReq->ifr_name,
`
As I am not familar with patch and so on I hope it is understandable. Could someone change this in source with next commit?

Thanks.

**Comment by mrjimenez** [comment]
Hi robolaf,

I did a small change in master today in these lines, but your change does not seem to be correct.

sizeof *pifReq == sizeof struct ifreq

sizeof ifConf.ifc_req == sizeof struct ifreq *

You have changed the increment to the size of a pointer. Since the variable i is a char index into the buffer (ifc_len is in bytes), that does not seem to be what we want.

Regards,
Marcelo.

**Comment by robolaf** [comment]
Hi Marcelo,
Thanks for the quick reply and cleaning up the code… You are right, of course. This was the wrong increment… 
So, I digged into it a little bit more and found the following on my system:
* the returned ifConf.ifc_len is 72
* I looked at all 72 bytes in the buffer and there are two interfaces: starting at byte 0:lo and at 36:br0
=> that means the correct increment of i for my system needs to be 36 here:
 ```
	/* Cycle through the list of interfaces looking for IP addresses.  */
	for (i = 0; i < ifConf.ifc_len; i +=  36) {     //(int)(sizeof (struct ifreq))) {
 ```
I have no glue whether this is the same on other systems or whether ioctl sends back wrong data in my case.
I hope you have the possibility to check this out… and fix it accordingly.

Thanks and regards.

**Comment by v00d00** [comment]
@robolaf Did you have success with @mrjimenez's patch? => https://github.com/mrjimenez/pupnp/commit/0fbbe083474221ef3c4484b33ecfe62a3faabba5

**Comment by robolaf** [comment]
@v00d00 No, as said. The patch helps to better read the code, however it did not help to solve the issue.

**Comment by eliasp** [comment]
Seeing the same issue on Gentoo, using version 1.10.1.
By default, Gerbera/libupnp binds to this local interface:
```
8: bridge0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 3a:54:61:0d:f4:75 brd ff:ff:ff:ff:ff:ff
    inet 10.27.1.62/24 brd 10.27.1.255 scope global dynamic bridge0
       valid_lft 301sec preferred_lft 301sec
    inet6 fe80::3854:61ff:fe0d:f475/64 scope link 
       valid_lft forever preferred_lft forever
```

But as soon as I try to bind it to the following interface, I'll end up with `LibUPnP error code: -121`:
```
10: wg-sae13: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none 
    inet 10.13.13.2/24 brd 10.13.13.255 scope global wg-sae13
       valid_lft forever preferred_lft forever
```

Is there anything I could provide to debug this further?

**Comment by mrjimenez** [comment]
Hi Elias,

The return code -121 is UPNP_E_INVALID_INTERFACE. If you search for this, you will see that this is only returned in function UpnpGetIfInfo() in upnpapi.c. This function is called from UpnpInit2(). That means that UpnpGetIfInfo() could not find the interface name you requested or that it does not have an IP address bound to it.

I would suggest that you put some printf's in UpnpGetIfInfo() to find out what is going on.

Regards,
Marcelo.

**Comment by Vollstrecker** [comment]
@eliasp Is this still an issue or is it gone with the (few) releases since then?

**Comment by mrjimenez** [comment]
Hi all,

The original SIOCGIFCONF increment issue robolaf found has since been replaced — the current code uses `getifaddrs()` instead of a manual `ifc_req` walk.

The remaining bug, confirmed by @eliasp, is that `UpnpGetIfInfo()` unconditionally filtered out any interface lacking `IFF_MULTICAST` — including WireGuard, IPIP tunnels, and other point-to-point interfaces — even when the caller explicitly named that interface via `UpnpInit2()`. The fix is to apply the `IFF_MULTICAST` filter only during auto-selection (`IfName == NULL`). When a specific interface is named, the library will now accept it; note that SSDP discovery will not function on non-multicast interfaces, but the rest of the UPnP stack will.

A fix and regression tests are available in PR #570.
